add python API to print all operators that have kernels registered to a particular DispatchKey

bdhirsh opened the pull request:
Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63575

Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

bdhirsh committed:
add python API to print all operators that have kernels registered to a particular DispatchKey

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30426919).

Chillee reviewed (approved):
LGTM!

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -29,16 +29,71 @@ torch::Library::Kind parseKind(const std::string& k) {
 
 c10::optional<c10::DispatchKey> parseDispatchKey(const std::string& k) {
   static std::unordered_map<std::string, c10::DispatchKey> key_map = {
+    {"", c10::DispatchKey::Undefined},
```
Need to make it easier to keep this up to date. Consider moving this in to c10/core/DispatchKey.cpp

ezyang reviewed (approved):
Thanks a lot; I be this will be useful. Needs a test.

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh commented:
I moved the DispatchKey parsing into c10 and added some tests. The tests are a little annoying - it looks like you can't use normal python stdout capture logic like `contextlib.redirect_stdout`, because it won't pick up stuff printed to stdout from C++. I added a util for capturing C++ stdout inspired from https://www.py4u.net/discuss/66399.

I also didn't want to use expect test, since the output will change with every new op. So I just tested the error cases, and that "[cpu ops] are a subset of [all ops] but not the other way around".

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30426919).

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30426919).

bdhirsh committed:
Update on "add python API to print all operators that have kernels registered to a particular DispatchKey"

Expose a Python API for printing all operator names that have a kernel registered to a particular DispatchKey. I talked to @Chillee and hopefully this will make the process of tracking the ops that a backend has coverage of less manual.

A better long-term thing would be to also calculate which `CompositeImplicitAutograd` ops a given backend inherits for free, based on the decomposition.

Example (not run with functorch, so only printing stuff registered in core):
```
>>> import torch
>>> torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')
aten::flatten.using_ints
aten::linalg_norm.ord_str
aten::quantile.new
...
* prints 911 operator names *
>>> torch._C._dispatch_print_registrations_for_dispatch_key('Batched')
* prints 130 operator names *
```

You'll get everything, not just aten ops. Specifically, there are quantized ops that are registered in core. It should be easy to filter / group the output as needed though.



Differential Revision: [D30426919](https://our.internmc.facebook.com/intern/diff/D30426919)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30426919).
